lib/Sugar.ml

```diff
@@ -291,49 +291,74 @@ let sequence (conf : Conf.t) cmts xexp =
+    ; pmty_loc
+    ; pmty_attributes }
+    when for_functor_kw || (List.is_empty pmty_attributes && is_valid arg.txt)
     ->
-      let arg =
-        if String.equal "*" arg.txt then {arg with txt= ""} else arg
-      in
       Cmts.relocate cmts ~src:pmty_loc ~before:arg.loc ~after:body.pmty_loc ;
       let body = sub_mty ~ctx body in
       let xargs, xbody =
         match pmty_attributes with
         | [] -> functor_type cmts ~for_functor_kw body
         | _ -> ([], body)
       in
-      ((arg, Option.map arg_mty ~f:(sub_mty ~ctx)) :: xargs, xbody)
+      ((arg, Some (sub_mty ~ctx arg_mty)) :: xargs, xbody)
+  | {pmty_desc= Pmty_functor (Unit, body); pmty_attributes; _}
+    when for_functor_kw || List.is_empty pmty_attributes ->
+      let body = sub_mty ~ctx body in
+      let xargs, xbody =
+        match pmty_attributes with
+        | [] -> functor_type cmts ~for_functor_kw body
+        | _ -> ([], body)
+      in
+      (({txt= Some ""; loc= Location.none}, None) :: xargs, xbody)
   | _ -> ([], xmty)
 
 (* The sugar is different when used with the [functor] keyword. The syntax
    M(A : A)(B : B) cannot handle [_] as module name. *)
 let rec functor_ cmts ~for_functor_kw ~source_is_long ({ast= me; _} as xme) =
-  let valid_sugared_name arg =
-    not (String.equal arg.txt "_" && source_is_long me)
+  let is_valid = function
+    | Some "_" -> false
+    | Some _ -> true
+    | None -> false
   in
   let ctx = Mod me in
   match me with
-  | {pmod_desc= Pmod_functor (arg, arg_mt, body); pmod_loc; pmod_attributes}
+  | { pmod_desc= Pmod_functor (Named (arg, arg_mt), body)
+    ; pmod_loc
+    ; pmod_attributes }
     when for_functor_kw
-         || (List.is_empty pmod_attributes && valid_sugared_name arg) ->
-      let arg =
-        if String.equal "*" arg.txt then {arg with txt= ""} else arg
-      in
+         || List.is_empty pmod_attributes
+            && not ((not (is_valid arg.txt)) && source_is_long me) ->
       Cmts.relocate cmts ~src:pmod_loc ~before:arg.loc ~after:body.pmod_loc ;
-      let xarg_mt = Option.map arg_mt ~f:(sub_mty ~ctx) in
+      let xarg_mt = sub_mty ~ctx arg_mt in
+      let ctx = Mod body in
+      let body = sub_mod ~ctx body in
+      let xargs, xbody_me =
+        match pmod_attributes with
+        | [] -> functor_ cmts ~for_functor_kw ~source_is_long body
+        | _ -> ([], body)
+      in
+      ((arg, Some xarg_mt) :: xargs, xbody_me)
+  | {pmod_desc= Pmod_functor (Unit, body); pmod_attributes; _}
+    when for_functor_kw
+         || (List.is_empty pmod_attributes && not (source_is_long me)) ->
       let ctx = Mod body in
       let body = sub_mod ~ctx body in
       let xargs, xbody_me =
         match pmod_attributes with
         | [] -> functor_ cmts ~for_functor_kw ~source_is_long body
         | _ -> ([], body)
       in
-      ((arg, xarg_mt) :: xargs, xbody_me)
+      (({txt= Some ""; loc= Location.none}, None) :: xargs, xbody_me)
```

Comment:
So `None` is for `(_)` and `Some ""` is `()` ? What about having a dedicated type for this ? (instead of `label option loc`, which have too many meanings)

Reply:
We shouldn't be afraid to change the whole function and use the types that make it safe instead of the types that make the code shorter.

Reply:
I packed the 2 values in the same option type, it looks better

Reply:
That's a nice improvement !

That wasn't my suggestion but now the type is very similar to the corresponding part of the parsetree, which is enough.